c/zis/client.c

```diff
@@ -162,25 +164,51 @@ int zisGenerateOrValidateSafIdt(const CrossMemoryServerName *serverName,
   }
   strncpy(parmList.passwordNullTerm, password, sizeof(parmList.passwordNullTerm));
 
-  parmList.options |= ZIS_AUTH_SERVICE_PARMLIST_OPTION_GENERATE_IDT;
-  parmList.safIdtLen = strlen(safIdt);
+  if (appl != NULL && strlen(appl) != 0) {
+    if (strlen(appl) >= sizeof (parmList.safIdtService.applNullTerm)) {
+      status->baseStatus.serviceRC = RC_ZIS_AUTHSRV_INPUT_STRING_TOO_LONG;
+      return RC_ZIS_SRVC_SERVICE_FAILED;
+    }
+    strcpy(parmList.safIdtService.applNullTerm, appl);
+
+    parmList.options |= ZIS_AUTH_SERVICE_PARMLIST_OPTION_IDT_APPL;
+  }
 
-  if (strlen(safIdt) >= sizeof(parmList.safIdt)) {
+  parmList.safIdtService.safIdtLen = strlen(safIdt);
+  parmList.safIdtService.safIdtServiceVersion = 1;
+  if (strlen(safIdt) >= sizeof(parmList.safIdtService.safIdt)) {
     status->baseStatus.serviceRC = RC_ZIS_AUTHSRV_INPUT_STRING_TOO_LONG;
     return RC_ZIS_SRVC_SERVICE_FAILED;
   }
-  memcpy((void *)parmList.safIdt, (void *)safIdt, strlen(safIdt));
+  memcpy((void *)parmList.safIdtService.safIdt, (void *)safIdt, strlen(safIdt));
 
   int rc = authRequest(serverName, &parmList, status);
 
-  if (parmList.safIdtLen > 0) {
+  if (parmList.safIdtService.safIdtLen > 0) {
     memset((void *)safIdt, 0, ZIS_AUTH_SERVICE_PARMLIST_SAFIDT_LENGTH + 1);
-    memcpy((void *)safIdt, (void *)parmList.safIdt, parmList.safIdtLen);
+    memcpy((void *)safIdt, (void *)parmList.safIdtService.safIdt, parmList.safIdtService.safIdtLen);
```

Comment:
Is the result SAF IDT a normal string without a null-terminator? Can `\0` be a valid character in a SAF IDT? If yes, how will the caller distinguish between that and the null-terminator?

I'm asking because the SAF call returns the length of the result, that is, it doesn't operate with C-strings.

Reply:
I don't know, this is the only documentation I found https://www.ibm.com/docs/en/zos/2.4.0?topic=reference-activating-using-idta-parameter-in-racroute-requestverify

Reply:
From the page:
> Identity Token Formats
> The Identity Token Area (IDTA) is a flexible interface designed to support different types of Identity Tokens (IDT) going forward. Currently RACROUTE has support for an IDT in the format of a JSON Web Token (JWT). **In the future RACROUTE may support other IDT formats**.

Since you only use JWT according to the code:
```c
idtaVar.idtType = IDTA_JWT_IDT_Type;
```
I suggest updating the doc for these public functions telling that it only support JWT. You will not be able to support anything else in the future because of the API you have here (relying on `\0`).